### ISSUE TITLE
SQLAlchemy 0.9.6 InvalidRequestError due to missing extend_existing=True in extensions' init_db()

### ISSUE THREAD

**Issue by florianm** [ckan/ckan#2679]
e.g. with a datacats "latest" ckan and ckanext-pages, installing the extensions results in:
```
...
(ckanext-pages gets called for the first time)
File '/project/ckanext-pages/ckanext/pages/actions.py', line 223 in pages_list
  return _pages_list(context, data_dict)
File '/project/ckanext-pages/ckanext/pages/actions.py', line 76 in _pages_list
  db.init_db(context['model'])
File '/project/ckanext-pages/ckanext/pages/db.py', line 112 in init_db
  sa.Column('extras', types.UnicodeText, default=u'{}'),
File '/usr/lib/ckan/local/lib/python2.7/site-packages/sqlalchemy/sql/schema.py', line 339 in __new__
  "existing Table object." % key)
InvalidRequestError: Table 'ckanext_pages' is already defined for this MetaData instance.  Specify 'extend_existing=True' to redefine options and columns on an existing Table object.
```

From the SQLAlchemy 0.9 [docs](http://docs.sqlalchemy.org/en/rel_0_9/core/metadata.html#sqlalchemy.schema.Table.params.extend_existing) I can see that the requested `extend_existing` parameter defaults to `False`.
SQLALchemy [version 0.7](http://docs.sqlalchemy.org/en/rel_0_8/changelog/migration_07.html?highlight=extend_existing#useexisting-superseded-by-extend-existing-and-keep-existing) changed:
```
The useexisting flag on Table has been superseded by a new pair of flags keep_existing and extend_existing. extend_existing is equivalent to useexisting - the existing Table is returned, and additional constructor elements are added. With keep_existing, the existing Table is returned, but additional constructor elements are not added - these elements are only applied when the Table is newly created.
```

However, ckan's SQLAlchemy was upgraded in [July](https://github.com/ckan/ckan/commit/81d64ec4b7f2b2993cee2a1c01640669de6e5704) from 0.7.8 to 0.9.6, and I've seen CKAN work fine since then.

So the code that immediately throws the error worked fine until I ran into datacats/datacats#334.
There's a chance this might be datacats related, there's a chance it's something silly I did on my end (plain vanilla CKAN plus a few extensions off the shelf), there's a chance this is some change in CKAN that requires an update to many extensions.

Anyone else has this problem yet?
I'll try to define `extend_existing` and see whether that fixes it. I'm too spoilt by Django's ORM that never requires me to talk to the db personally that I'm pretty new to SQLAlchemistry.
Could some more experienced alchemist point me into the right direction to troubleshoot?

It seems that adding "extend_existing=True" to each sa.Column() inside an extension's `init_db()` fixes that issue. I'm running into other issues ATM and can't confirm this fixes the problem yet.

**Comment by amercader** [comment]
@florianm Just to clarify and before we spend time on it, is this still an issue? Is saw some comments on the datacats issue that suggest it might not be

**Comment by joetsoi** [comment]
so what I suspect is happening is the way ckanext-pages handles it's database, [init_db()](https://github.com/ckan/ckanext-pages/blob/09d229b0021b66eb0bbb10185b1af91b02073989/ckanext/pages/db.py#L20) is called multiple times by multiple actions ([see here](https://github.com/ckan/ckanext-pages/search?utf8=%E2%9C%93&q=init_db) this would cause the pages tables to be mapped several times to sqlalchemy metadata causing this error. stuff like ckanext-harvest table setup is more well behaved and only runs once at start up which is probably why we haven't bumped into this until now.

**Comment by florianm** [comment]
@amercader @joetsoi I've got one instance back working without the outdated SQLAlchemy==0.7.8 and with datacats/datacats#334 solved, and another where a stray SQLAlchemy==0.7.8 still gets pulled in somehow. Will double-check whether I'm using upstream or own form of ckanext-pages and try and reproduce above issue now.

**Comment by florianm** [comment]
Good news, this issue with `extend_existing` seems to be gone now that datacats/datacats#334 is fixed.

A plain vanilla datacats ckan (version "latest") with ckanext-pages [works](http://test.alpha.data.wa.gov.au/pages/example), has a [datapusher](http://test.alpha.data.wa.gov.au/api/action/status_show) and SQLAlchemy==0.9.6 as it should have.

I'm closing this issue as resolved. Thanks to you all for fixing issues up the chain of datapusher and ckan-service-provider, which ultimately unblocked datacats/datacats#334.

**Comment by florianm** [comment]
Reopening with interesting finding.
I've installed additional extensions (ckanext-pages is upstream/master/latest) into the test instance I reported working above, and am running into the problem with pages again.
* List of extensions: http://test.alpha.data.wa.gov.au/api/action/status_show
* repos used and install steps: 
http://govhack2015.readthedocs.org/en/latest/ckan.html#extensions
 * did not run this step http://govhack2015.readthedocs.org/en/latest/ckan.html#manage-dependency-conclicts
* initialised db (spatial, harvester) without error message (spatial is my fork with `extend_existing`)
* pip freeze in the environment:
```
APScheduler==2.1.2
Babel==0.9.6
Beaker==1.7.0
Flask==0.9
Flask-Admin==1.3.0
Flask-Login==0.3.0
FormEncode==1.2.6
Genshi==0.6
GeoAlchemy==0.7.2
GeoAlchemy2==0.2.5
Jinja2==2.6
Mako==0.9.0
MarkupSafe==0.18
OWSLib==0.8.6
Pairtree==0.7.1-T
Paste==1.7.5.1
PasteDeploy==1.5.0
PasteScript==1.7.5
Pygments==1.6
Pylons==0.9.7
Routes==1.13
SPARQLWrapper==1.6.4
SQLAlchemy==0.9.6
Shapely==1.3.1
Tempita==0.5.2
Unidecode==0.04.18
WTForms==2.0.2
WebError==0.10.3
WebHelpers==1.3
WebOb==1.0.8
WebTest==1.4.3
Werkzeug==0.10.4
amqplib==1.0.2
anyjson==0.3.3
argparse==1.2.1
celery==2.4.2
chardet==2.3.0
-e git+https://github.com/ckan/ckan@85108d89ebff59fcbf4ee847b8e234c93329bc85#egg=ckan-master
-e git+https://github.com/ckan/ckanapi.git@341c5a56e9f40f6c3a58fe00a108423de569879c#egg=ckanapi-master
json-table-schema==0.2.1
kombu==2.1.3
kombu-sqlalchemy==1.1.0
lxml==3.3.5
messytables==0.15.0
nose==1.3.0
ofs==0.4.1
passlib==1.6.2
pbr==0.8.2
pika==0.9.8
psycopg2==2.4.5
-e git+https://github.com/geopython/pycsw.git@7adcf58226c39a05784666e23e8c50ad6823bf38#egg=pycsw-master
pyparsing==1.5.6
pyproj==1.9.3
pyshp==1.2.3
python-dateutil==2.4.2
python-magic==0.4.6
python-slugify==1.1.4
pytz==2012j
pyutilib.component.core==4.5.3
rdflib==4.1.2
rdflib-jsonld==0.2
redis==2.10.1
repoze.lru==0.6
repoze.who==2.0
repoze.who-friendlyform==1.0.8
requests==2.8.1
simplejson==3.3.1
six==1.7.3
solrpy==0.9.5
sqlalchemy-migrate==0.9.1
sqlparse==0.1.11
tzlocal==1.2
unicodecsv==0.9.4
vdm==0.13
wsgiref==0.1.2
xlrd==0.9.4
xmltodict==0.9.0
zope.interface==4.1.1
```

Note: SQLAlchemy is at 0.9.8 out of the box.
Reloading datacats did work, but the GUI shows an error trace back to ckanext-pages' db.py reported at ckan/ckanext-pages#27
I suspect if I add `extend_existing=True` to db.py, `datacats install` and `datacats reload` everything will be running fine again.

**Comment by florianm** [comment]
Fix is here: https://github.com/ckan/ckanext-pages/pull/28
